src/Microsoft.ML.DataView/DataViewTypeManager.cs

```diff
@@ -0,0 +1,247 @@
+        /// </summary>
+        private class TypeWithAttributesId
+        {
+            public Type TargetType { get; }
+            private IEnumerable<Attribute> _associatedAttributes;
+
+            public TypeWithAttributesId(Type rawType, IEnumerable<Attribute> attributes)
+            {
+                TargetType = rawType;
+                _associatedAttributes = attributes;
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (obj is TypeWithAttributesId other)
+                {
+                    // Flag of having the same type.
+                    var sameType = TargetType.Equals(other.TargetType);
+                    // Flag of having the attribute configurations.
+                    var sameAttributeConfig = true;
+
+                    if (_associatedAttributes == null && other._associatedAttributes == null)
+                        sameAttributeConfig = true;
+                    else if (_associatedAttributes == null && other._associatedAttributes != null)
+                        sameAttributeConfig = false;
+                    else if (_associatedAttributes != null && other._associatedAttributes == null)
+                        sameAttributeConfig = false;
+                    else if (_associatedAttributes.Count() != other._associatedAttributes.Count())
+                        sameAttributeConfig = false;
+                    else
+                    {
+                        var zipped = _associatedAttributes.Zip(other._associatedAttributes, (attr, otherAttr) => (attr, otherAttr));
+                        foreach (var (attr, otherAttr) in zipped)
+                        {
+                            if (!attr.Equals(otherAttr))
+                                sameAttributeConfig = false;
+                        }
+                    }
+
+                    return sameType && sameAttributeConfig;
+                }
+                return false;
+            }
+
+            /// <summary>
+            /// This function computes a hashing ID from <see name="TargetType"/> and attributes attached to it.
+            /// If a type is defined as a member in a <see langword="class"/>, <see name="Attributes"/> can be obtained by calling
+            /// <see cref="MemberInfo.GetCustomAttributes(bool)"/>.
+            /// </summary>
+            public override int GetHashCode()
+            {
+                if (_associatedAttributes == null)
+                    return TargetType.GetHashCode();
+
+                var code = TargetType.GetHashCode();
+                foreach (var attr in _associatedAttributes)
+                    code = Hashing.CombineHash(code, attr.GetHashCode());
+                return code;
+            }
+
+        }
+
+        /// <summary>
+        /// An instance of <see cref="DataViewTypeId"/> represents an unique key of its <see cref="TargetType"/>.
+        /// </summary>
+        private class DataViewTypeId
```

Comment:
Is this class necessary? It just wraps `DataViewType` and relies on its underlying Equals and GetHashCode. I think instead, we could just use `DataViewType` as the key to the dictionary.

Reply:
Nice catch. I will remove it.